src/storage/data_table.cpp

```diff
@@ -437,7 +437,112 @@ static void VerifyCheckConstraint(TableCatalogEntry &table, Expression &expr, Da
 	}
 }
 
-void DataTable::VerifyAppendConstraints(TableCatalogEntry &table, DataChunk &chunk) {
+static bool FindColumnIndex(const vector<idx_t> *keys_ptr, const vector<column_t> &index_column_ids) {
+	for (idx_t i = 0; i < keys_ptr->size(); i++) {
+		bool is_found = false;
+		for (idx_t j = 0; j < index_column_ids.size(); j++) {
+			if ((*keys_ptr)[i] == index_column_ids[j]) {
+				is_found = true;
+				break;
+			}
+		}
+		if (!is_found) {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static void VerifyForeignKeyConstraint(const BoundForeignKeyConstraint &bfk, ClientContext &context, DataChunk &chunk,
+                                       bool is_append) {
+	const vector<idx_t> *src_keys_ptr = &bfk.info.fk_keys;
+	const vector<idx_t> *dst_keys_ptr = &bfk.info.pk_keys;
+	if (!is_append) {
+		src_keys_ptr = &bfk.info.pk_keys;
+		dst_keys_ptr = &bfk.info.fk_keys;
+	}
+
+	auto table_entry_ptr =
+	    Catalog::GetCatalog(context).GetEntry<TableCatalogEntry>(context, bfk.info.schema, bfk.info.table);
+	if (table_entry_ptr == nullptr) {
+		throw InternalException("Can't find table \"%s\" in foreign key constraint", bfk.info.table);
+	}
+
+	// make the data chunk to check
+	vector<LogicalType> types;
+	for (idx_t i = 0; i < table_entry_ptr->columns.size(); i++) {
+		types.emplace_back(table_entry_ptr->columns[i].type);
+	}
+	DataChunk dst_chunk;
+	dst_chunk.InitializeEmpty(types);
+	for (idx_t i = 0; i < src_keys_ptr->size(); i++) {
+		dst_chunk.data[(*dst_keys_ptr)[i]].Reference(chunk.data[(*src_keys_ptr)[i]]);
+	}
+	dst_chunk.SetCardinality(chunk.size());
+	auto data_table = table_entry_ptr->storage.get();
+
+	idx_t count = dst_chunk.size();
+	if (count <= 0) {
+		return;
+	}
+
+	vector<string> err_msgs, tran_err_msgs;
```

Comment:
This array is need because we must check both data table's index and transaction local storage's index.
You can see this by looking at the code below using these arrays.

Reply:
My point is that this does not need to be an array, it can be a single `string` (`string err_msg`) - since only ever one error message string is used. This should simplify a bunch of the code here, and also make it significantly faster.

Reply:
But we need to look concurrently the error messages in data table's index and transaction local storage's index.
Please look up at the following code.

	for (idx_t i = 0; i < count; i++) {
		if ((!transaction_check && !err_msgs[i].empty()) ||
		    (transaction_check && ((is_append && !err_msgs[i].empty() && !tran_err_msgs[i].empty()) ||
		                           (!is_append && (!err_msgs[i].empty() || !tran_err_msgs[i].empty()))))) {
			string &err_msg = err_msgs[i];
			if (err_msg.empty()) {
				err_msg = tran_err_msgs[i];
			}
			throw ConstraintException(err_msg);
		}
	}